crates/runtime/src/libcalls.rs

```diff
@@ -434,50 +438,89 @@ unsafe fn externref_global_set(vmctx: *mut VMContext, index: u32, externref: *mu
     memory_index: u32,
-    addr: u64,
-    _expected: u32,
-    _timeout: u64,
+    addr_index: u64,
+    expected: u32,
+    timeout: u64,
 ) -> Result<u32, TrapReason> {
+    let timeout = (timeout as i64 >= 0).then(|| Duration::from_nanos(timeout));
+
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 4, 4)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_wait32) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let (addr, spot) = validate_atomic_addr(instance, memory, addr_index, 4, 4)?;
+
+    // SAFETY: `addr` was validated by `validate_atomic_addr` above.
+    let atomic = unsafe { &*(addr as *const AtomicU32) };
+
+    // We want the sequential consistency of `SeqCst` to ensure that the `load` sees the value that the `notify` will/would see.
+    // All WASM atomic operations are also `SeqCst`.
+    let validate = || atomic.load(Ordering::SeqCst) == expected;
+
+    do_atomic_wait(addr_index, timeout, spot, validate)
 }
 
 // Implementation of `memory.atomic.wait64` for locally defined memories.
 unsafe fn memory_atomic_wait64(
     vmctx: *mut VMContext,
     memory_index: u32,
-    addr: u64,
-    _expected: u64,
-    _timeout: u64,
+    addr_index: u64,
+    expected: u64,
+    timeout: u64,
 ) -> Result<u32, TrapReason> {
+    let timeout = (timeout as i64 >= 0).then(|| Duration::from_nanos(timeout));
+
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 8, 8)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_wait64) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let (addr, spot) = validate_atomic_addr(instance, memory, addr_index, 8, 8)?;
+
+    // SAFETY: `addr` was validated by `validate_atomic_addr` above.
+    let atomic = unsafe { &*(addr as *const AtomicU64) };
+
+    // We want the sequential consistency of `SeqCst` to ensure that the `load` sees the value that the `notify` will/would see.
+    // All WASM atomic operations are also `SeqCst`.
+    let validate = || atomic.load(Ordering::SeqCst) == expected;
+
+    do_atomic_wait(addr_index, timeout, spot, validate)
+}
+
+#[inline]
+fn do_atomic_wait(
+    addr_index: u64,
+    timeout: Option<Duration>,
+    spot: Option<Arc<ParkingSpot>>,
+    validate: impl FnOnce() -> bool,
+) -> Result<u32, TrapReason> {
+    if let Some(spot) = spot {
+        match spot.park(addr_index, validate, timeout) {
+            ParkResult::Unparked => Ok(0),
+            ParkResult::Invalid => Ok(1),
+            ParkResult::TimedOut => Ok(2),
+        }
+    } else if let Some(timeout) = timeout {
+        std::thread::sleep(timeout);
+        Ok(2)
+    } else {
+        Err(TrapReason::Wasm(TrapCode::AlwaysTrapAdapter))
+    }
```

Comment:
It appears I didn't read close enough, and [according to the overview](https://github.com/WebAssembly/threads/blob/main/proposals/threads/Overview.md) (which I would highly recommend reviewing/correlating with the implementation here) it says:

> Wait operators additionally trap if used on an unshared linear memory

So this can't actually sleep it needs to always trap. Additionally you'll want to allocate a trap code instead of reusing a different trap code for this purpose.